Source/ProgressionSystemRuntime/Private/Data/PSWorldSubsystem.cpp

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) Valerii Rotermel & Yevhenii Selivanov
+
+#include "Data/PSWorldSubsystem.h"
+
+#include "Components/PSHUDComponent.h"
+#include "Controllers/MyPlayerController.h"
+#include "Data/PSDataAsset.h"
+#include "Data/PSSaveGameData.h"
+#include "Kismet/GameplayStatics.h"
+#include "LevelActors/PlayerCharacter.h"
+#include "MyDataTable/MyDataTable.h"
+#include "MyUtilsLibraries/UtilsLibrary.h"
+#include "Engine/Engine.h"
+#include "Engine/World.h"
+#include "MyUtilsLibraries/GameplayUtilsLibrary.h"
+#include "Subsystems/GlobalEventsSubsystem.h"
+#include "UtilityLibraries/MyBlueprintFunctionLibrary.h"
+
+#include UE_INLINE_GENERATED_CPP_BY_NAME(PSWorldSubsystem)
+
+// Returns this Subsystem, is checked and wil crash if can't be obtained
+UPSWorldSubsystem& UPSWorldSubsystem::Get()
+{
+	const UWorld* World = UUtilsLibrary::GetPlayWorld();
+	checkf(World, TEXT("%s: 'World' is null"), *FString(__FUNCTION__));
+	UPSWorldSubsystem* ThisSubsystem = World->GetSubsystem<ThisClass>();
+	checkf(ThisSubsystem, TEXT("%s: 'ProgressionSubsystem' is null"), *FString(__FUNCTION__));
+	return *ThisSubsystem;
+}
+
+// Returns this Subsystem, is checked and wil crash if can't be obtained
+UPSWorldSubsystem& UPSWorldSubsystem::Get(const UObject& WorldContextObject)
+{
+	const UWorld* World = GEngine->GetWorldFromContextObjectChecked(&WorldContextObject);
+	checkf(World, TEXT("%s: 'World' is null"), *FString(__FUNCTION__));
+	UPSWorldSubsystem* ThisSubsystem = World->GetSubsystem<ThisClass>();
+	checkf(ThisSubsystem, TEXT("%s: 'ProgressionSubsystem' is null"), *FString(__FUNCTION__));
+	return *ThisSubsystem;
+}
+// Returns current row of progression system
+const FPSRowData& UPSWorldSubsystem::GetCurrentRow() const
+{
+	const UPSSaveGameData* SaveGameInstance = GetCurrentSaveGameData();
+	checkf(SaveGameInstance, TEXT("ERROR: 'SaveGameInstanceInternal' is null"));
+
+	return SaveGameInstance->GetCurrentRow();
+}
+
+// Returns the data asset that contains all the assets of Progression System game feature
+const UPSDataAsset* UPSWorldSubsystem::GetPSDataAsset() const
+{
+	return PSDataAssetInternal.LoadSynchronous();
+}
+
+// Set the progression system component
+void UPSWorldSubsystem::SetHUDComponent(UPSHUDComponent* MyHUDComponent)
+{
+	checkf(MyHUDComponent, TEXT("%s: My progression system component is null"), *FString(__FUNCTION__));
+	PSHUDComponentInternal = MyHUDComponent;
+}
+
+// Called when world is ready to start gameplay before the game mode transitions to the correct state and call BeginPlay on all actors 
+void UPSWorldSubsystem::OnWorldBeginPlay(UWorld& InWorld)
+{
+	Super::OnWorldBeginPlay(InWorld);
+
+	// Subscribe events on player type changed and Character spawned
+	BIND_ON_LOCAL_CHARACTER_READY(this, ThisClass::UPSWorldSubsystem::OnCharacterReady);
```

Comment:
Please replace `ThisClass::UPSWorldSubsystem::OnCharacterReady` by `ThisClass::OnCharacterReady`. I wonder how is it even complies:D